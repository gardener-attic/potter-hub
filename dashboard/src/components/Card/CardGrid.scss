.CardGrid {
  display: flex;
  margin: 0 -1%;
  justify-content: space-between;
  flex-wrap: wrap;
}

.CardGrid:after {
  /* Display the last row properly: https://stackoverflow.com/a/34816625 */
  flex: auto;
  content: "";
}
