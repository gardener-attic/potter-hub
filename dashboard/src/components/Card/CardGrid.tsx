import * as React from "react";

import "./CardGrid.css";

export interface ICardGridProps {
  className?: string;
  children?: React.ReactChildren | React.ReactNode | string;
}

interface ICardGridState {
  scrollEndIndex: number
}

const SCROLL_THRESHOLD_TOP_PIXEL = 120;
const SCROLL_THRESHOLD_BOTTOM_TO_LOAD = 0.8;
const ITEMS_PER_SCROLL = 32

class CardGrid extends React.Component<ICardGridProps, ICardGridState> {
  constructor(props: any) {
    super(props);
    this.state = {
      scrollEndIndex: ITEMS_PER_SCROLL,
    };
  }

  public componentDidMount() {
    window.addEventListener("scroll", this.onScroll);
  }

  public componentWillUnmount() {
    window.removeEventListener("scroll", this.onScroll)
  }

  public render() {
    return (
      <div className={`CardGrid padding-v-big ${this.props.className || ""}`}>
        {React.Children.toArray(this.props.children).slice(0, this.state.scrollEndIndex)}
      </div>
    );
  }

  private onScroll = (): void => {
    // calculates, if the current scroll position further than the SCROLL_THRESHOLD_BOTTOM_TO_LOAD threshold
    const isScrollAtBottom = window.innerHeight + document.documentElement.scrollTop >
      document.documentElement.offsetHeight * SCROLL_THRESHOLD_BOTTOM_TO_LOAD;

    // calculates, if the current scroll position is above the SCROLL_THRESHOLD_TOP_PIXEL
    const isScrollAtTop = document.documentElement.scrollTop < SCROLL_THRESHOLD_TOP_PIXEL;

    if (isScrollAtBottom && this.state.scrollEndIndex < React.Children.count(this.props.children)) {
      this.setState((state) => ({
        scrollEndIndex: state.scrollEndIndex + ITEMS_PER_SCROLL
      }))
    }

    if (isScrollAtTop && this.state.scrollEndIndex > ITEMS_PER_SCROLL) {
      this.setState((state) => ({
        scrollEndIndex: ITEMS_PER_SCROLL
      }))
    }
  }
}

export default CardGrid;
