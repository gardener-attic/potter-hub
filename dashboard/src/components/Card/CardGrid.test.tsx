import { shallow } from "enzyme";
import * as React from "react";

import CardGrid from "./CardGrid";

it("should render the className", () => {
  const wrapper = shallow(<CardGrid className="foo" />);
  expect(wrapper.find(".foo")).toExist();
});

it("should render the children elements", () => {
  const wrapper = shallow(
    <CardGrid>
      <div>foo</div>
    </CardGrid>,
  );
  expect(wrapper.text()).toContain("foo");
  expect(wrapper).toMatchSnapshot();
});
