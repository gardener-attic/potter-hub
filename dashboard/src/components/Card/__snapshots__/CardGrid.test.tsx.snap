// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the children elements 1`] = `
<div
  className="CardGrid padding-v-big "
>
  <div
    key=".0"
  >
    foo
  </div>
</div>
`;
